components/tidb_query/src/rpn_expr/impl_string.rs

```diff
@@ -93,6 +93,27 @@ pub fn rtrim(arg: &Option<Bytes>) -> Result<Option<Bytes>> {
     }))
 }
 
+#[rpn_fn]
+#[inline]
+pub fn replace(
+    s: &Option<Bytes>,
+    from_str: &Option<Bytes>,
+    to_str: &Option<Bytes>,
+) -> Result<Option<Bytes>> {
+    Ok(match (s, from_str, to_str) {
+        (Some(s), Some(from_str), Some(to_str)) => {
+            if from_str.is_empty() {
+                return Ok(Some(s.clone()));
+            }
+            let s = String::from_utf8_lossy(&s);
+            let from_str = String::from_utf8_lossy(&from_str);
+            let to_str = String::from_utf8_lossy(&to_str);
```

Comment:
I think we should use `from_utf8` and shortcut the `Err` by `?`

Reply:
```rust
let invalid = b"Hello \xF0\x90\x80World";
assert!(String::from_utf8(invalid.to_vec()).is_err()); // got error
assert_eq!("Hello �World", String::from_utf8_lossy(invalid));
```
I tested the invalid characters in play.tidb.io and mysql8 respectively. no errors occurred.
```
TiDB> select replace('Hello �World', 'World', '123');
+-------------------------------------------+
| replace("Hello �World", 'World', '123')   | 
+-------------------------------------------+
| Hello �123                                | 
+-------------------------------------------+
1 row in set (0.04 sec)
```
```
mysql> select replace('Hello �World', 'World', '123');
+-------------------------------------------+
| replace('Hello �World', 'World', '123')   |
+-------------------------------------------+
| Hello �123                                |
+-------------------------------------------+
1 row in set (0.01 sec)
```

Reply:
You're right; no error is thrown. But there is a misbehavior on `from_utf8_lossy` that it replaces the invalid utf8 chars:

```sql
mysql> select HEX(replace(X'F09080', '', ''));
+---------------------------------+
| HEX(replace(X'F09080', '', '')) |
+---------------------------------+
| F09080                          |
+---------------------------------+
1 row in set (0.00 sec)
```

```rust
let bytes = vec![0xF0,0x90,0x80,0x00,0x61];
let str = String::from_utf8_lossy(&bytes);
assert_eq!(str.as_bytes(), &bytes[..])
```

```text
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `[239, 191, 189, 0, 97]`,
 right: `[240, 144, 128, 0, 97]`', src/main.rs:4:5
```

Reply:
you can use this lib to implement the replacement on binary level: https://crates.io/crates/twoway

Reply:
Hi! I found this in the document [rust String doc](https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8_lossy).

> If you are sure that the byte slice is valid UTF-8, and you don't want to incur the overhead of the conversion, there is an unsafe version of this function, from_utf8_unchecked, which has the same behavior but skips the checks.
```rust 
let bytes = vec![0xF0,0x90,0x80,0x00,0x61];
let s = unsafe { str::from_utf8_unchecked(&bytes) };
assert_eq!(s.as_bytes(), &bytes[..]);
```
That Looks ok. Can i use it?

Reply:
> If you are sure that the byte slice is valid UTF-8, ...

But it's not. A str containing invaid chars will cause undefined behavior.

Reply:
Yes! use two-way instead, please. It's faster.

Reply:
Ok, Think you very much!